Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch

```diff
@@ -0,0 +1,131 @@
+From 996f744b89e05a99e74d5768d13363cfb8975a41 Mon Sep 17 00:00:00 2001
+From: Sam Ezeh <spammy23@users.noreply.github.com>
+Date: Wed, 30 Aug 2017 20:34:26 +0100
+Subject: [PATCH] Implemented the cancellable ban events
+
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java b/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java
+index 80832f78..05791767 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java
+@@ -1,19 +1,19 @@
+ package org.bukkit.craftbukkit;
+ 
+-import java.io.IOException;
+-import java.net.InetSocketAddress;
+-import java.util.Date;
+-import java.util.Set;
+-
++import com.google.common.collect.ImmutableSet;
+ import net.minecraft.server.IpBanEntry;
+ import net.minecraft.server.IpBanList;
+-import net.minecraft.server.MinecraftServer;
+ import org.apache.commons.lang.StringUtils;
+ import org.apache.commons.lang.Validate;
++import org.bukkit.Bukkit;
++import org.bukkit.craftbukkit.event.CraftEventFactory;
++import org.bukkit.event.server.BanIpEvent;
+ 
+-import com.google.common.collect.ImmutableSet;
++import java.io.IOException;
++import java.net.InetSocketAddress;
++import java.util.Date;
++import java.util.Set;
+ import java.util.logging.Level;
+-import org.bukkit.Bukkit;
+ 
+ public class CraftIpBanList implements org.bukkit.BanList {
+     private final IpBanList list;
+@@ -36,13 +36,20 @@ public class CraftIpBanList implements org.bukkit.BanList {
+ 
+     @Override
+     public org.bukkit.BanEntry addBan(String target, String reason, Date expires, String source) {
++
+         Validate.notNull(target, "Ban target cannot be null");
+ 
+         IpBanEntry entry = new IpBanEntry(target, new Date(),
+                 StringUtils.isBlank(source) ? null : source, expires,
+                 StringUtils.isBlank(reason) ? null : reason);
+ 
+-        list.add(entry);
++        BanIpEvent event = new BanIpEvent(new CraftIpBanEntry(target, entry, list));
++        event = CraftEventFactory.callEvent(event);
++
++        if(event.isCancelled()) return null;
```

Comment:
This API is pretty much contracted to not return null, do change that contract now would potentially be breaking